File: truckdoc-client-m2/application/src/main/java/com/sanda/truckdoc/client/data/MessagesDatabaseService.java
package com.sanda.truckdoc.client.data;

import android.content.Context;

import com.j256.ormlite.android.apptools.OpenHelperManager;
import com.j256.ormlite.dao.Dao;
import com.j256.ormlite.dao.RuntimeExceptionDao;
import com.j256.ormlite.misc.TransactionManager;
import com.j256.ormlite.stmt.QueryBuilder;
import com.j256.ormlite.stmt.Where;
import com.j256.ormlite.table.TableUtils;
import com.sanda.truckdoc.client.api.AttachmentPojo;
import com.sanda.truckdoc.client.api.ClientToServerMessagePojo;
import com.sanda.truckdoc.client.api.v2.ServerToClientMessagePojoNew;
import com.sanda.truckdoc.client.api.v3.sync.routing.model.NamedPoint;
import com.sanda.truckdoc.client.data.model.AttachmentInfo;
import com.sanda.truckdoc.client.data.model.DbContactRecord;
import com.sanda.truckdoc.client.data.model.DbLocation;
import com.sanda.truckdoc.client.data.model.MessageFileRecord;
import com.sanda.truckdoc.client.data.model.ServerMessage;
import com.sanda.truckdoc.client.data.model.route.DbRouteAssignment;
import com.sanda.truckdoc.client.data.model.route.DbRoutePath;
import com.sanda.truckdoc.client.data.model.route.DbRoutePoint;
import com.sanda.truckdoc.client.util.FileHelper;
import com.sanda.truckdoc.client.util.RoleTypeMapper;
import com.sanda.truckdoc.client.util.UnexpectedException;
import com.sanda.truckdoc.client.util.commons.FileUtils;

import org.joda.time.DateTime;

import java.io.File;
import java.io.IOException;
import java.sql.SQLException;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.inject.Inject;
import javax.inject.Singleton;

import rx.Observable;
import rx.schedulers.Schedulers;
import timber.log.Timber;

import static rx.Observable.from;
import static rx.Observable.just;

/**
 * @author Alexei Osipov
 */
@Singleton
public class MessagesDatabaseService {

    private final RuntimeExceptionDao<ServerMessage, Integer> serverMessagesDao;
    private final RuntimeExceptionDao<AttachmentInfo, Integer> attachmentDao;
    private final RuntimeExceptionDao<DbLocation, Long> locationDao;
    private final RuntimeExceptionDao<DbContactRecord, Long> contactRecordDao;
    private final RuntimeExceptionDao<MessageFileRecord, Long> messageFileDao;
    private final RuntimeExceptionDao<DbRouteAssignment, Long> routeAssignmentDao;
    private final RuntimeExceptionDao<DbRoutePoint, Long> routePointDao;
    private final RuntimeExceptionDao<DbRoutePath, Long> routePathDao;


    @Inject
    public MessagesDatabaseService(DatabaseHelper helper) {
        serverMessagesDao = helper.getRuntimeExceptionDao(ServerMessage.class);
        attachmentDao = helper.getRuntimeExceptionDao(AttachmentInfo.class);
        locationDao = helper.getRuntimeExceptionDao(DbLocation.class);
        contactRecordDao = helper.getRuntimeExceptionDao(DbContactRecord.class);
        messageFileDao = helper.getRuntimeExceptionDao(MessageFileRecord.class);
        routeAssignmentDao = helper.getRuntimeExceptionDao(DbRouteAssignment.class);
        routePointDao = helper.getRuntimeExceptionDao(DbRoutePoint.class);
        routePathDao = helper.getRuntimeExceptionDao(DbRoutePath.class);
    }

    public void initializeRoutePoints(DbRoutePath route, List<NamedPoint> points) {
        routePathDao.assignEmptyForeignCollection(route, "points");
        for (NamedPoint point : points) {
            DbRoutePoint dbRoutePoint = new DbRoutePoint(point);
            dbRoutePoint.setPath(route);
            routePointDao.create(dbRoutePoint);
        }
        routePathDao.update(route);
    }

    public Observable<Integer> markMessageAsDownloaded(ServerMessage message) {
        return Observable.defer(() -> {
            message.setDownloaded(true);
            return just(serverMessagesDao.update(message));
        });
    }

    public Observable<Integer> insertLocation(DbLocation location) {
        return Observable.defer(() -> {
            return just(locationDao.create(location));
        });
    }

    public DbRouteAssignment insertRouteAssignment(DbRouteAssignment routeAssignment) {
        return routeAssignmentDao.createIfNotExists(routeAssignment);
    }

    public List<DbRouteAssignment> findRouteAssignmentByServerId(Long serverRouteAssignmentId) {
        return routeAssignmentDao.queryForEq("serverAssignmentId", serverRouteAssignmentId);
    }

    public DbRouteAssignment findRouteAssignmentById(Long routeAssignmentId) {
        return routeAssignmentDao.queryForId(routeAssignmentId);
    }


    public void addSmsMessage(ServerMessage serverMessage) {
        serverMessagesDao.create(serverMessage);
    }

    public void markHidden(ServerMessage serverMessage) {
        serverMessage.setHidden(true);
        serverMessagesDao.update(serverMessage);
    }

    public List<DbLocation> getLocations() {
        try {
            return locationDao.queryBuilder().limit(100L).query();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public Observable<DbContactRecord> getContactRecords() {
        return Observable.defer(() -> from(contactRecordDao.queryForAll()));
    }

    public Observable<MessageFileRecord> getMessageFiles() {
        return Observable.fromCallable(messageFileDao::queryForAll).flatMap(Observable::from);
    }

    public Observable<MessageFileRecord> getNotSentMessageFiles() {
        return Observable.fromCallable(() -> messageFileDao.queryForEq("sent", false)).flatMap(Observable::from);
    }

    public Observable<MessageFileRecord> getNotUploadedMessageFiles() {
        return Observable.fromCallable(() -> messageFileDao.queryBuilder().where().isNull("serverId").query()).flatMap(Observable::from);
    }

    public void createMessageFileRecord(MessageFileRecord record) {
        messageFileDao.create(record);
    }

    public void deleteMessageFileRecords(List<MessageFileRecord> records) {
        for (MessageFileRecord record : records) {
            new File(record.getPath()).delete();
        }
        messageFileDao.delete(records);
    }

    public void deleteAllMessageFileRecords() {
        List<MessageFileRecord> records = messageFileDao.queryForAll();
        for (MessageFileRecord record : records) {
            File f = new File(record.getPath());
            if (f.exists()) f.delete();
        }
        messageFileDao.delete(records);
    }

    public void updateMessageFileRecord(MessageFileRecord record) {
        messageFileDao.update(record);
    }

    public boolean replaceContactRecords(List<DbContactRecord> newContacts) {
        try {
            TransactionManager.callInTransaction(contactRecordDao.getConnectionSource(), () -> {
                TableUtils.clearTable(contactRecordDao.getConnectionSource(), DbContactRecord.class);
                for (DbContactRecord contact : newContacts) {
                    contactRecordDao.create(contact);
                }
                return null;
            });
            return true;
        } catch (SQLException e) {
            Timber.e(e, "Replace Contact Records failed");
            return false;
        }
    }

    public Observable<Integer> updateContactRecord(DbContactRecord record) {
        return Observable.defer(() -> just(contactRecordDao.update(record)));
    }

    public void deleteAssignments() {
        for (DbRouteAssignment dbRouteAssignment : routeAssignmentDao.queryForAll()) {
            routePointDao.delete(dbRouteAssignment.getRoute().getPoints());
            routePathDao.delete(dbRouteAssignment.getRoute());
            routeAssignmentDao.delete(dbRouteAssignment);
        }
    }

    public void deleteLocations(List<DbLocation> locations) {
        locationDao.delete(locations);
    }

    public Observable<Integer> markFileAsDownloaded(final AttachmentInfo attachment) {
        return Observable.defer(() -> {
            attachment.setDownloaded(true);
            return just(attachmentDao.update(attachment));
        });
    }

    public Observable<Integer> deleteAllMessages() {
        return Observable.defer(() -> {
            QueryBuilder<ServerMessage, Integer> qb = serverMessagesDao.queryBuilder();
            qb.orderBy("saved", true);
            List<ServerMessage> list = null;
            try {
                list = serverMessagesDao.query(qb.prepare());
            } catch (SQLException e) {
                throw new RuntimeException(e);
            }
            for (ServerMessage message : list) {
                try {
                    FileUtils.deleteDirectory(FileHelper.getIncomeDirectory(message.getId(), "*"));
                } catch (IOException e) {
                    //
                }
                serverMessagesDao.delete(message);
            }
            return just(0);
        }).subscribeOn(Schedulers.newThread());
    }

    //called in bg
    public static List<ServerMessage> getNotDownloadedMessages(Context context) {
        return new DatabaseOperation<List<ServerMessage>, ServerMessage>() {
            @Override
            public List<ServerMessage> impl(Dao<ServerMessage, Integer> dao) throws SQLException {
                return dao.queryForEq("downloaded", false);
            }
        }.execute(context, ServerMessage.class);
    }

    //called in bg
    public static List<AttachmentInfo> getNotDownloadedFiles(Context context, final ServerMessage message) {
        return new DatabaseOperation<List<AttachmentInfo>, AttachmentInfo>() {
            @Override
            public List<AttachmentInfo> impl(Dao<AttachmentInfo, Integer> dao) throws SQLException {
                QueryBuilder<AttachmentInfo, Integer> qb = dao.queryBuilder();
                return dao.query(qb.where().eq("message_id", message.getId()).and().eq("downloaded", false).prepare());
            }
        }.execute(context, AttachmentInfo.class);
    }

    public static void saveMessageIfNotExist(final Context context, final ServerToClientMessagePojoNew messagePojo) {
        new DatabaseOperation<Void, ServerMessage>() {
            @Override
            public Void impl(Dao<ServerMessage, Integer> dao) throws SQLException {
                List<ServerMessage> oldMessages = dao.queryForEq("serverMessageId", messagePojo.getId());
                if (oldMessages.size() == 0) {
                    ServerMessage message = new ServerMessage(messagePojo);
                    message.setSavedDate(DateTime.now());
                    message.setServerMessageId(messagePojo.getId());

                    if (messagePojo.getAttachments().size() > 0) {
                        message.setDownloaded(false);
                    }
                    dao.create(message);
                    saveFiles(context, messagePojo.getAttachments(), message);
                }
                return null;
            }
        }.execute(context, ServerMessage.class);
    }

    public static void saveOutgoingMessages(final Context context,
                                            final Collection<ClientToServerMessagePojo> messagePojos) {
        new DatabaseOperation<Void, ServerMessage>() {
            @Override
            public Void impl(Dao<ServerMessage, Integer> dao) throws SQLException {
                for (ClientToServerMessagePojo messagePojo : messagePojos) {
                    ServerMessage oldMessage = null;
                    if (messagePojo.getId() != null) {
                        oldMessage = dao.queryForId(messagePojo.getId());
                        oldMessage.setSent(messagePojo.isSent());
                        dao.update(oldMessage);
                    }
                    if (oldMessage == null) {
                        ServerMessage message = new ServerMessage(messagePojo);
                        message.setSavedDate(DateTime.now());
                        message.setOutgoing(true);
                        message.setSent(messagePojo.isSent());
                        message.setProcessed(false);
                        message.setDownloaded(true);
                        message.setRecipientId(messagePojo.getRecipientId());
                        dao.create(message);
                        saveFiles(context, messagePojo.getAttachments(), message);
                    }
                }
                return null;
            }
        }.execute(context, ServerMessage.class);
    }

    public static void saveFiles(Context context,
                                 final Collection<AttachmentPojo> attachmentPojos,
                                 final ServerMessage message) {
        new DatabaseOperation<Void, AttachmentInfo>() {
            @Override
            public Void impl(Dao<AttachmentInfo, Integer> dao) throws SQLException {
                for (AttachmentPojo attachmentPojo : attachmentPojos) {
                    AttachmentInfo attachmentInfo = new AttachmentInfo(attachmentPojo);
                    attachmentInfo.setMessage(message);
                    dao.create(attachmentInfo);
                }
                return null;
            }
        }.execute(context, AttachmentInfo.class);
    }

    public Observable<ServerMessage> getMessages(boolean showHidden) {
        //for debug
        ServerMessage sm = new ServerMessage();
        sm.setDownloaded(true);
        sm.setSavedDate(DateTime.now());
        sm.setRecipientId(RoleTypeMapper.IN_COLUMN);
        sm.setText("text");
        sm.setSenderRoleId(RoleTypeMapper.OUT_EXPEDITER);
        sm.setOutgoing(false);
        sm.setId(888000);

        return Observable.defer(() -> {
            //SELECT * FROM server_message WHERE (outgoing=1 OR (outgoing=0 AND downloaded=1)) AND hidden?=0
            QueryBuilder<ServerMessage, Integer> q = serverMessagesDao.queryBuilder();
            Where<ServerMessage, Integer> where = q.where();
            try {
                where.or(where.eq("outgoing", true), where //
                        .eq("outgoing", false).and().eq("downloaded", true));
                if (!showHidden) {
                    where.and(where, where.eq("hidden", showHidden));
                }
                q.orderBy("savedDate", false);
                q.limit(50L);
                return from(serverMessagesDao.query(q.prepare()));
            } catch (SQLException e) {
                throw new RuntimeException(e);
            }
        });
    }

    public static List<ServerMessage> getPendingOutMessages(Context context) {
        return new DatabaseOperation<List<ServerMessage>, ServerMessage>() {
            @Override
            public List<ServerMessage> impl(Dao<ServerMessage, Integer> dao) throws SQLException {
                Map<String, Object> map = new HashMap<>();
                map.put("sent", false);
                map.put("outgoing", true);
                return dao.queryForFieldValuesArgs(map);
            }
        }.execute(context, ServerMessage.class);
    }

    public static void deleteOldMessages(Context context) {
        new DatabaseOperation<Void, ServerMessage>() {
            @Override
            public Void impl(Dao<ServerMessage, Integer> dao) throws SQLException {

                QueryBuilder<ServerMessage, Integer> qb = dao.queryBuilder();
                qb.orderBy("saved", true);
                List<ServerMessage> list = dao.query(qb.prepare());
                for (ServerMessage message : list) {
                    DateTime date = message.getSavedDate();
                    try {
                        if (date.isBefore(DateTime.now().minusMonths(2))) {
                            FileUtils.deleteDirectory(FileHelper.getIncomeDirectory(message.getId(), "*"));
                            dao.delete(message);
                        }
                    } catch (IOException e) {
                        //
                    }
                }
                return null;
            }
        }.execute(context, ServerMessage.class);
    }

    public void deleteAllData() {
        try {
            FileHelper.deleteAllAppFiles();
            TableUtils.clearTable(attachmentDao.getConnectionSource(), attachmentDao.getDataClass());
            TableUtils.clearTable(serverMessagesDao.getConnectionSource(), serverMessagesDao.getDataClass());
            TableUtils.clearTable(locationDao.getConnectionSource(), routeAssignmentDao.getDataClass());
            TableUtils.clearTable(contactRecordDao.getConnectionSource(), contactRecordDao.getDataClass());
            TableUtils.clearTable(messageFileDao.getConnectionSource(), messageFileDao.getDataClass());
            TableUtils.clearTable(routeAssignmentDao.getConnectionSource(), routeAssignmentDao.getDataClass());
            TableUtils.clearTable(routePointDao.getConnectionSource(), routePointDao.getDataClass());
            TableUtils.clearTable(routePathDao.getConnectionSource(), routePathDao.getDataClass());
        } catch (Exception e) {
            //throw new RuntimeException(e);
        }

    }

    private static abstract class DatabaseOperation<R, T> {

        public abstract R impl(Dao<T, Integer> dao) throws SQLException;

        public R execute(Context context, Class<T> entityClass) {
            DatabaseHelper helper = OpenHelperManager.getHelper(context, DatabaseHelper.class);
            try {
                @SuppressWarnings("unchecked") Dao<T, Integer> dao = helper.getDao(entityClass);
                return impl(dao);
            } catch (SQLException e) {
                throw new UnexpectedException("Failed to execute database operation", e);
            } finally {
                OpenHelperManager.releaseHelper();
            }
        }
    }
}

File: truckdoc-client-m2/application/src/main/java/com/sanda/truckdoc/client/data/MessagesDatabaseService.kt
package com.sanda.truckdoc.client.data

import com.sanda.truckdoc.client.api.AttachmentPojo
import com.sanda.truckdoc.client.api.ClientToServerMessagePojo
import com.sanda.truckdoc.client.api.v2.ServerToClientMessagePojoNew
import com.sanda.truckdoc.client.api.v3.sync.routing.model.NamedPoint
import com.sanda.truckdoc.client.data.dao.*
import com.sanda.truckdoc.client.data.model.*
import com.sanda.truckdoc.client.data.model.route.*
import com.sanda.truckdoc.client.util.FileHelper
import com.sanda.truckdoc.client.util.RoleTypeMapper
import com.sanda.truckdoc.client.util.commons.FileUtils
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.withContext
import org.joda.time.DateTime
import timber.log.Timber
import java.io.File
import java.io.IOException
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class MessagesDatabaseService @Inject constructor(
    private val serverMessageDao: ServerMessageDao,
    private val attachmentDao: AttachmentDao,
    private val locationDao: LocationDao,
    private val contactRecordDao: ContactRecordDao,
    private val messageFileDao: MessageFileDao,
    private val routeAssignmentDao: RouteAssignmentDao,
    private val routePointDao: RoutePointDao,
    private val routePathDao: RoutePathDao
) {
    // Static methods for Java compatibility
    companion object {
        suspend fun deleteOldMessages(context: android.content.Context) {
            // This would need to be implemented with proper dependency injection
            Timber.w("deleteOldMessages called but not implemented")
        }
        
        suspend fun saveOutgoingMessages(context: android.content.Context, messages: List<ClientToServerMessagePojo>) {
            // This would need to be implemented with proper dependency injection
            Timber.w("saveOutgoingMessages called but not implemented")
        }
        
        suspend fun saveMessageIfNotExist(context: android.content.Context, message: ServerToClientMessagePojoNew) {
            // This would need to be implemented with proper dependency injection
            Timber.w("saveMessageIfNotExist called but not implemented")
        }
        
        suspend fun getNotDownloadedMessages(context: android.content.Context): List<ServerMessage> {
            // This would need to be implemented with proper dependency injection
            Timber.w("getNotDownloadedMessages called but not implemented")
            return emptyList()
        }
        
        suspend fun getNotDownloadedFiles(context: android.content.Context, message: ServerMessage): List<AttachmentInfo> {
            // This would need to be implemented with proper dependency injection
            Timber.w("getNotDownloadedFiles called but not implemented")
            return emptyList()
        }
    }

    suspend fun deleteAllData() {
        withContext(Dispatchers.IO) {
            deleteAllMessages()
            deleteAllMessageFileRecords()
            contactRecordDao.deleteAll()
            locationDao.deleteAll()
            routeAssignmentDao.deleteAll()
            routePathDao.deleteAll()
            routePointDao.deleteAll()
        }
    }

    suspend fun initializeRoutePoints(routeId: Long, points: List<NamedPoint>) {
        withContext(Dispatchers.IO) {
            for ((index, point) in points.withIndex()) {
                val routePoint = DbRoutePoint(
                    id = 0,
                    path_id = routeId,
                    name = point.name ?: "Point ${index + 1}",
                    latitude = point.lat,
                    longitude = point.lng,
                    order = index
                )
                routePointDao.insert(routePoint)
            }
        }
    }

    suspend fun markMessageAsDownloaded(message: ServerMessage): Int {
        return withContext(Dispatchers.IO) {
            val updatedMessage = message.copy(downloaded = true)
            serverMessageDao.update(updatedMessage)
        }
    }

    suspend fun insertLocation(location: DbLocation): Long {
        return withContext(Dispatchers.IO) {
            locationDao.insert(location)
        }
    }

    suspend fun insertRouteAssignment(routeAssignment: DbRouteAssignment): Long {
        return withContext(Dispatchers.IO) {
            routeAssignmentDao.insert(routeAssignment)
        }
    }

    fun findRouteAssignmentByServerId(serverRouteAssignmentId: Long): Flow<List<DbRouteAssignment>> {
        return routeAssignmentDao.findByServerId(serverRouteAssignmentId)
    }

    fun findRouteAssignmentById(routeAssignmentId: Long): Flow<DbRouteAssignment?> {
        return routeAssignmentDao.findById(routeAssignmentId)
    }

    suspend fun addSmsMessage(serverMessage: ServerMessage) {
        withContext(Dispatchers.IO) {
            serverMessageDao.insert(serverMessage)
        }
    }

    suspend fun markHidden(serverMessage: ServerMessage) {
        withContext(Dispatchers.IO) {
            val updatedMessage = serverMessage.copy(hidden = true)
            serverMessageDao.update(updatedMessage)
        }
    }

    fun getLocations(): Flow<List<DbLocation>> {
        return locationDao.getAll()
    }

    fun getContactRecords(): Flow<List<DbContactRecord>> {
        return contactRecordDao.getAll()
    }

    fun getMessageFiles(): Flow<List<MessageFileRecord>> {
        return messageFileDao.getAll()
    }

    fun getNotSentMessageFiles(): Flow<List<MessageFileRecord>> {
        return messageFileDao.getNotSent()
    }

    fun getNotUploadedMessageFiles(): Flow<List<MessageFileRecord>> {
        return messageFileDao.getNotUploaded()
    }

    suspend fun createMessageFileRecord(record: MessageFileRecord) {
        withContext(Dispatchers.IO) {
            messageFileDao.insert(record)
        }
    }

    suspend fun deleteMessageFileRecords(records: List<MessageFileRecord>) {
        withContext(Dispatchers.IO) {
            for (record in records) {
                File(record.path).delete()
            }
            messageFileDao.delete(records)
        }
    }

    suspend fun deleteAllMessageFileRecords() {
        withContext(Dispatchers.IO) {
            val records = messageFileDao.getAllSync()
            for (record in records) {
                File(record.path).delete()
            }
            messageFileDao.deleteAll()
        }
    }

    suspend fun updateMessageFileRecord(record: MessageFileRecord) {
        withContext(Dispatchers.IO) {
            messageFileDao.update(record)
        }
    }

    suspend fun replaceContactRecords(newContacts: List<DbContactRecord>): Boolean {
        return withContext(Dispatchers.IO) {
            try {
                contactRecordDao.deleteAll()
                contactRecordDao.insertAll(newContacts)
                true
            } catch (e: Exception) {
                Timber.e(e, "Replace Contact Records failed")
                false
            }
        }
    }

    suspend fun updateContactRecord(record: DbContactRecord): Int {
        return withContext(Dispatchers.IO) {
            contactRecordDao.update(record)
        }
    }

    suspend fun deleteAssignments() {
        withContext(Dispatchers.IO) {
            val assignments = routeAssignmentDao.getAllSync()
            for (assignment in assignments) {
                routePointDao.deleteByPathId(assignment.route_id)
                val route = routePathDao.findById(assignment.route_id).first()
                route?.let { routePathDao.delete(it) }
                routeAssignmentDao.delete(assignment)
            }
        }
    }

    suspend fun deleteLocations(locations: List<DbLocation>) {
        withContext(Dispatchers.IO) {
            locationDao.delete(locations)
        }
    }

    suspend fun markFileAsDownloaded(attachment: AttachmentInfo): Int {
        return withContext(Dispatchers.IO) {
            val updatedAttachment = attachment.copy(downloaded = true)
            attachmentDao.update(updatedAttachment)
        }
    }

    suspend fun deleteAllMessages() {
        withContext(Dispatchers.IO) {
            val messages = serverMessageDao.getAllSync()
            for (message in messages) {
                try {
                    FileUtils.deleteDirectory(FileHelper.getIncomeDirectory(message.id, "*"))
                } catch (e: IOException) {
                    Timber.e(e)
                }
                serverMessageDao.delete(message)
            }
        }
    }

    fun getMessages(showHidden: Boolean): Flow<List<ServerMessage>> {
        return if (showHidden) {
            serverMessageDao.getAll()
        } else {
            serverMessageDao.getNotHidden()
        }
    }

    fun getPendingOutMessages(): Flow<List<ServerMessage>> {
        return serverMessageDao.getPendingOutMessages()
    }

    suspend fun deleteOldMessages() {
        withContext(Dispatchers.IO) {
            val messages = serverMessageDao.getAllSync()
            val twoMonthsAgo = DateTime.now().minusMonths(2)
            
            for (message in messages) {
                if (message.savedDate.isBefore(twoMonthsAgo)) {
                    try {
                        FileUtils.deleteDirectory(FileHelper.getIncomeDirectory(message.id, "*"))
                        serverMessageDao.delete(message)
                    } catch (e: IOException) {
                        Timber.e(e)
                    }
                }
            }
        }
    }

    suspend fun saveOutgoingMessages(messages: List<ClientToServerMessagePojo>) {
        withContext(Dispatchers.IO) {
            for (message in messages) {
                val serverMessage = ServerMessage(
                    id = 0,
                    message = message.text,
                    savedDate = DateTime.now(),
                    recipientId = message.recipientId?.toLong(),
                    senderUserId = null,
                    senderVirtualGroupId = null,
                    downloaded = true,
                    hidden = false,
                    outgoing = true,
                    sent = message.isSent(),
                    attachments = emptyList()
                )
                serverMessageDao.insert(serverMessage)
            }
        }
    }

    suspend fun saveMessageIfNotExist(message: ServerToClientMessagePojoNew) {
        withContext(Dispatchers.IO) {
            val existingMessage = serverMessageDao.getByServerId(message.id)
            if (existingMessage.isEmpty()) {
                val serverMessage = ServerMessage(
                    id = 0,
                    serverMessageId = message.id,
                    message = message.text,
                    savedDate = DateTime.now(),
                    recipientId = null, // This field doesn't exist in the message
                    senderUserId = message.senderUserId,
                    senderVirtualGroupId = message.senderVirtualGroupId,
                    downloaded = false,
                    hidden = false,
                    outgoing = false,
                    sent = false,
                    attachments = emptyList()
                )
                val messageId = serverMessageDao.insert(serverMessage)
                
                // Insert attachments
                for (attachment in message.attachments) {
                    val attachmentInfo = AttachmentInfo(
                        id = 0,
                        message_id = messageId.toInt(),
                        name = attachment.fileName,
                        downloaded = false
                    )
                    attachmentDao.insert(attachmentInfo)
                }
            }
        }
    }

    suspend fun getNotDownloadedMessages(): List<ServerMessage> {
        return withContext(Dispatchers.IO) {
            val allMessages = serverMessageDao.getAllSync()
            allMessages.filter { !it.downloaded }
        }
    }

    suspend fun getNotDownloadedFiles(message: ServerMessage): List<AttachmentInfo> {
        return withContext(Dispatchers.IO) {
            attachmentDao.getNotDownloadedByMessageId(message.id)
        }
    }

    suspend fun updateMessageFile(file: MessageFileRecord): Int {
        return withContext(Dispatchers.IO) {
            messageFileDao.update(file)
        }
    }
} 